File: app/src/main/java/ua/syt0r/furiganit/viewmodel/history/HistoryViewModel.java
package ua.syt0r.furiganit.viewmodel.history;

import android.app.Application;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import java.util.List;

import io.reactivex.Completable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import ua.syt0r.furiganit.model.entity.HistoryItem;
import ua.syt0r.furiganit.model.repository.hisotry.HistoryRepository;
import ua.syt0r.furiganit.model.repository.hisotry.firestore.FirestoreHistoryRepository;
import ua.syt0r.furiganit.model.repository.hisotry.local.LocalHistoryRepository;
import ua.syt0r.furiganit.model.repository.user.UserRepository;

public class HistoryViewModel extends AndroidViewModel {

    private MutableLiveData<List<HistoryItem>> mutableHistory = new MutableLiveData<>();

    private MutableLiveData<Throwable> mutableError = new MutableLiveData<>();
    private MutableLiveData<Boolean> shouldAskSignIn = new MutableLiveData<>();

    private UserRepository userRepository;
    private HistoryRepository localHistoryRepository;
    private HistoryRepository remoteHistoryRepository;

    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    public HistoryViewModel(Application application) {
        super(application);

        userRepository = new UserRepository();
        localHistoryRepository = new LocalHistoryRepository(application);

        if (userRepository.isLogged())
            remoteHistoryRepository = new FirestoreHistoryRepository();

    }

    public void removeItems(int position) {

        List<HistoryItem> history = mutableHistory.getValue();
        if (history != null) {

            HistoryItem item = history.get(position);

            Disposable disposable = localHistoryRepository.remove(item)
                    .andThen(remoteHistoryRepository != null ?
                            remoteHistoryRepository.remove(item) : Completable.complete())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(
                            () -> {

                            },
                            error -> {

                            });

            compositeDisposable.add(disposable);
        }

    }

    public void loadHistory() {

    }

    public void syncHistory() {

    }

    public LiveData<List<HistoryItem>> subscribeOnHistory() {
        return mutableHistory;
    }

    public LiveData<Boolean> subscribeOnSignInOption() {
        return shouldAskSignIn;
    }

    public LiveData<Throwable> subscribeOnError() {
        return mutableError;
    }

}

File: app/src/main/java/ua/syt0r/furiganit/app/history/HistoryViewModel.kt
package ua.syt0r.furiganit.app.history

import androidx.lifecycle.*
import com.google.firebase.auth.FirebaseAuth
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import ua.syt0r.furiganit.model.entity.HistoryItem
import ua.syt0r.furiganit.model.repository.hisotry.remote.RemoteHistoryRepository
import ua.syt0r.furiganit.model.repository.hisotry.local.LocalHistoryRepository
import ua.syt0r.furiganit.model.repository.hisotry.remote.RepoState
import ua.syt0r.furiganit.model.repository.hisotry.remote.SyncAction
import ua.syt0r.furiganit.model.repository.hisotry.remote.SyncState
import ua.syt0r.furiganit.model.usecase.TextLocalizerUseCase
import ua.syt0r.furiganit.utils.SingleLiveEvent
import java.lang.IllegalStateException

class HistoryViewModel(
        private val localHistoryRepository: LocalHistoryRepository,
        private val remoteHistoryRepository: RemoteHistoryRepository,
        private val firebaseAuth: FirebaseAuth,
        private val textLocalizerUseCase: TextLocalizerUseCase
) : ViewModel() {

    private val mutableHistory = MutableLiveData<List<HistoryItem>>()
    private val mutableSyncState = MutableLiveData<SyncState>()
    private val mutableMessage = SingleLiveEvent<String>()
    private val mutableEvent = SingleLiveEvent<Event>()

    private val compositeDisposable = CompositeDisposable()

    fun fetchHistory() {

        val disposable = localHistoryRepository
                .fetchHistory()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { mutableHistory.value = it },
                        { mutableMessage.value = textLocalizerUseCase.getErrorMessage(it.message) }
                )

        compositeDisposable.add(disposable)

    }

    fun checkRemoteRepoState() {

        if (firebaseAuth.currentUser == null) {
            mutableEvent.value = Event.GO_TO_SIGNUP_SCREEN
            return
        }

        val disposable = remoteHistoryRepository
                .checkRemoteRepo()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        { repoState ->
                            when(repoState) {
                                RepoState.NEWER -> { sync(SyncAction.OVERWRITE_LOCAL) }
                                RepoState.OUTDATED -> { sync(SyncAction.OVERWRITE_REMOTE) }
                                RepoState.DIFFERENT -> {
                                    mutableEvent.value = Event.SELECT_SYNC_ACTION
                                }
                                null -> throw IllegalStateException("Unknown repoState")
                            }
                        },
                        { mutableMessage.value = textLocalizerUseCase.getErrorMessage(it.message) }
                )

        compositeDisposable.add(disposable)

    }

    fun sync(syncAction: SyncAction) {

        val disposable = remoteHistoryRepository
                .sync(mutableHistory.value ?: listOf(), syncAction)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe {

                    mutableSyncState.value = when(syncAction) {
                        SyncAction.OVERWRITE_LOCAL -> SyncState.DOWNLOADING_CHANGES
                        SyncAction.OVERWRITE_REMOTE -> SyncState.UPLOADING_CHANGES
                        SyncAction.MERGE -> SyncState.MERGING
                    }

                }
                .subscribe(
                        {  },
                        { mutableMessage.value = textLocalizerUseCase.getErrorMessage(it.message) }
                )

        compositeDisposable.add(disposable)

    }


    fun subscribeOnHistory(): LiveData<List<HistoryItem>> {
        return mutableHistory
    }

    fun subscribeOnMessage(): LiveData<String> {
        return mutableMessage
    }

    fun subscribeOnEvents(): LiveData<Event> {
        return mutableEvent
    }

    override fun onCleared() {
        super.onCleared()
        compositeDisposable.dispose()
    }

    enum class Event { SELECT_SYNC_ACTION, GO_TO_SIGNUP_SCREEN }

}
